### PR TITLE
Android: Add support for flags and extras in StartActivity

### PR DESCRIPTION
**Pull request by a1rwulf**
## Description
Add support for extras and flags when starting android activities.

## Motivation and context
I'm planning to use this in an upcoming addon.

## How has this been tested?
Works as expected on my Nvidia Shield.

## What is the effect on users?
Addon developers can set flags and extras when running android intents.

## Screenshots (if appropriate):
None.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (howie-f) [inline xbmc/platform/android/activity/XBMCApp.cpp:973]: flags/extras now also default to ""

**Review by howie-f — APPROVED** [review]
i don't see anything wrong with this but can't test android things

**Review by ksooo — COMMENTED** [review]
Not sure how flags are supposed to work with the current implementation. ;-)

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (howie-f) [reply xbmc/platform/android/activity/XBMCApp.cpp:1009]: i was of the opinion `flags` will end up in `pMsg->params[4]` and intentFlags is only for validation and debuglog

https://github.com/xbmc/xbmc/blob/d1aefb88ef099739ddbc4ee90c4c10aaeb413460/xbmc/Application.cpp#L1515

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (lrusak) [reply xbmc/platform/android/activity/XBMCApp.cpp:983]: We do this elsewhere, but I'd like to get away from multiline log statements I think.

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp:983]: No problem, I'll either do separate log calls or put it in one line.

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (ksooo) [reply xbmc/platform/android/activity/XBMCApp.cpp:1009]: Why should `flags` be passed to this message, if nothing is done with it but checking the string can be converted to an int?

Comment (a1rwulf) [reply xbmc/platform/android/activity/XBMCApp.cpp:1009]: Somehow my comment got lost.
You're obviously right, I'll fix that 😉

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (ksooo) [inline xbmc/platform/android/activity/XBMCApp.cpp:1012]: I would not spend an extra var for this. Why not just `newIntent.setFlags(std::stoi(flags));`? If you would like to keep the var, make it const, please (`const int ...`).

**Review by ksooo — APPROVED** [review]

**Comment by a1rwulf** [comment]
Guess I'll need to change code formatting once again, because I forgot about it -.-
Will do 1 more force-push once the build is done and do another runtime test.
Thanks for your reviews!

**Comment by zach-morris** [comment]
Curious if this has been tested with Android 11+?  I believe the updated permissions/security cause issues (likely regardless of the use of these extras and flags).  I've run into the issue when attempting to launch external processes via an addon.  Reference [here](https://stackoverflow.com/questions/67648647/android-11-starting-a-service-of-another-app) and [here](https://developer.android.com/training/package-visibility).

**Comment by a1rwulf** [comment]
@zach-morris this change simply exposes 2 additional parameters for launching intents via the `StartActivity` API method.
I haven't experienced issues with launching activities on my Shield TV yet, which is Android 11 - however I've seen the error you've mentioned in https://github.com/zach-morris/plugin.program.iagl/issues/255 with other usecases.
It seemingly depends on wether the application to be launched, supports being launched: https://github.com/mmatyas/pegasus-frontend/issues/766#issuecomment-903820051

Long story, short - unfortunately I don't think it will magically make your stuff work, as it doesn't do anything regarding permissions etc. and you can't just bypass Androids restrictions by using intent-flags or intent-extras.

**Comment by a1rwulf** [comment]
Btw: Small correction:
The Shield is AndroidTV 11 which is not the same thing as Android 11 in some areas.
So actually no, it hasn't been tested on Android 11.

**Comment by chrisism** [comment]
This looks promising as also a solution to my item on [https://forum.kodi.tv/showthread.php?tid=367878](https://forum.kodi.tv/showthread.php?tid=367878).
How can we try it out?

**Comment by a1rwulf** [comment]
As noted, I have to fix an issue with setting flags first.
Then I can link a testbuild.

**Comment by chrisism** [comment]
> As noted, I have to fix an issue with setting flags first. Then I can link a testbuild.

Thanks, would be really nice.
Some more questions, looking at the code. What is the format of the flags and extras values? Since it is common with the commands to have multiple -e parameters, how would we supply those values?
Also I think currently it only accepts a package instead of the package/activity input. For example in the case of Retroarch now you can only supply "com.retroarch" where "com.retroarch/com.retroarch.browser.retroactivity.RetroActivityFuture" would be preferred.

**Comment by a1rwulf** [comment]
@chrisism I know our simple call does not match what you can do with the Intent API in the java framework and I don't want to spend too much time on it.
As for the intent: Is `com.retroarch.browser.retroactivity.RetroActivityFuture` your launch intent?
If you look at the code: https://github.com/xbmc/xbmc/blob/master/xbmc/platform/android/activity/XBMCApp.cpp#L981 Kodi queries the package manager for an intent to launch.
Regarding the extra data:
I think the easiest solution would be to make the extra parameter a json string similar to what the adbkit project does, so you could use sth like this:
```
[
  {
    key: "key1",
    type: "string",
    value: "val1"
  },
  {
    key: "key2",
    type: "string",
    value: "val2"
  },
  {
    key: "key3",
    type: "bool",
    value: true
  }
]

```

But I'm not sure yet, need to spend some more time thinking about it.

**Comment by chrisism** [comment]
@a1rwulf I believe the intent is actually "android.intent.action.MAIN".
It's something like this what we want to translate:
```
am start -n com.retroarch/.browser.retroactivity.RetroActivityFuture -a android.intent.action.MAIN -c android.intent.category.LAUNCHER -e ROM "xxxx" -e LIBRETRO xxxx-e CONFIGFILE xxxx
```
I tried running the StartAndroidActivity before with the activity in the package field, but that didnt work. Don't know if there is an alternative way of setting it.

And thanks, it is now clear about the extra params. That should work.

**Comment by a1rwulf** [comment]
@chrisism support for specifying an intent is already there, it's the second parameter:
https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/builtins/AndroidBuiltins.cpp

So something like this should work in current master:
```
xbmc.executebuiltin('StartAndroidActivity("com.retroarch","android.intent.action.MAIN")')
```

**Comment by lrusak** [comment]
jenkins build this please

**Comment by a1rwulf** [comment]
@chrisism Current state is untested and I won't have access to an android box for the next few days, but if you want you can give it a shot: https://jenkins.kodi.tv/job/Android-ARM/26220/artifact/kodi-20220730-95476bdf-PR21652-26220-armeabi-v7a.apk

**Comment by chrisism** [comment]
@a1rwulf Nice thanks. I'll give it a try here.

**Comment by chrisism** [comment]
Did a couple of tests. Succeeded in starting up a package without any extra parameters. Still busy on that part, getting the syntax correct. Won't have access to test box for a couple of days, so will continue later on.

**Comment by a1rwulf** [comment]
@chrisism I think the easiest way to execute your call is something like this:

Your example:
```
am start -n com.retroarch/.browser.retroactivity.RetroActivityFuture -a android.intent.action.MAIN -c android.intent.category.LAUNCHER -e ROM "xxxx" -e LIBRETRO xxxx-e CONFIGFILE xxxx```

```
            extradata = [
                {
                    "key": "ROM",
                    "type": "string",
                    "value": "xxxx"
                },
                {
                    "key": "LIBRETRO",
                    "type": "string",
                    "value": "xxxx"
                },
                {
                    "key": "CONFIGFILE",
                    "type": "string",
                    "value": "xxxx"
                }
            ]

            xbmc.executebuiltin('StartAndroidActivity("com.retroarch","","","", "", "{}")'.format(json.dumps(extradata)))
```

However I'm not sure if the current version will work.
There is no way to specify intent categories or actions as of now.

**Comment by a1rwulf** [comment]
My own usecase works fine with this version and I really want to get this into Nexus, so it would be good to get it merged in alpha stage.
I'm very well aware that the intent api of android has much more capabilities that might be useful, but that would probably require a redesign.
So I think further improvements are not necessarily subject to this PR.

File: xbmc/platform/android/activity/XBMCApp.cpp
Comment (howie-f) [inline xbmc/platform/android/activity/XBMCApp.cpp:1012]: maybe the whole block can (later) be changed to `std::from_chars`

**Review by howie-f — COMMENTED** [review]
can‘t contribute too much to android nor can i test it. i have to hand over to ksooo, but don‘t see anything that‘s holding this up

**Comment by chrisism** [comment]
@a1rwulf, Getting a bit further. Intent value is still bugging me though. Flags seems to be okay. I didnt escape quotes in the json string for the extras parameter, but it still picks it up. Didn't expect that.

Setting category and action might be indeed what is missing now for me. See this thread https://stackoverflow.com/questions/71984199/android-retroarch-load-rom-from-intent.

My log output now:
```
2022-08-14 17:09:30.923 T:14043   DEBUG <general>: StartActivity: package: com.retroarch
2022-08-14 17:09:30.924 T:14043   DEBUG <general>: StartActivity: intent: android.intent.action.MAIN
2022-08-14 17:09:30.924 T:14043   DEBUG <general>: StartActivity: dataType: 
2022-08-14 17:09:30.924 T:14043   DEBUG <general>: StartActivity: dataURI: 
2022-08-14 17:09:30.924 T:14043   DEBUG <general>: StartActivity: flags: 4096
2022-08-14 17:09:30.924 T:14043   DEBUG <general>: StartActivity: extras: [{"key": "ROM", "value": "'/storage/emulated/0/Download/work/nes/games/1943", "type": "string"}, {"key": "LIBRETRO", "value": "/data/user/0/com.retroarch/cores/nestopia_libretro_android.so", "type": "string"}, {"key": "CONFIGFILE", "value": "/storage/emulated/0/Android/data/com.retroarch/files/retroarch.cfg", "type": "string"}]
2022-08-14 17:09:30.944 T:14043   ERROR <general>: StartActivity: ExceptionOccurred launching com.retroarch
```

**Comment by a1rwulf** [comment]
> 

I had another look at the android intent docs and I think your call is correct now.

Intent has an constructor that takes an action:
https://developer.android.com/reference/android/content/Intent#Intent(java.lang.String)

And we do this in the code:
```
  CJNIIntent newIntent = intent.empty() ?
    GetPackageManager().getLaunchIntentForPackage(package) :
    CJNIIntent(intent);
```
Which means if you specify the second parameter it actually is an action not an intent name...
So my log with "intent: " is inaccurate.

I think it's rather easy to implement support to specifiy component and action, but the call will get slightly messy because of the number of parameters.

**Comment by chrisism** [comment]
@a1rwulf not to confuse but indeed I was trying to provide the correct value for the intent parameter from StartAndroidActivity() that translates into the intent action value that gets injected into the constructor. To do a more thorough check on my side, what is the actual format of the Intent action values that is expected? I had to convert the flag enum values to integers to get that one to work. So how about the Intent.ACTION_MAIN for example. Do you have some sample code or tests somewhere available for me to have a look at.

The action value from setAction() and perhaps addCategory are the methods that I am missing now to complete my own solution I think. I might have a look at that part myself. You didnt implement that in this PR so it is maybe something for a new branch/pr. It shouldn't block this PR for completion at least.

**Pull request merged**
